src/shader.h

```diff
@@ -74,6 +76,51 @@ class IShaderConstantSetter
 			bool is_highlevel) = 0;
 };
 
+
+class IShaderConstantSetterFactory
+{
+public:
+	virtual ~IShaderConstantSetterFactory() {};
+	virtual IShaderConstantSetter* create() = 0;
+};
+
+
+template <typename T, bool is_pixel, std::size_t count=1>
+class CachedShaderSetting {
+	const char *m_name;
+	T m_sent[count];
+	bool has_been_set;
+protected:
+	CachedShaderSetting(const char *name) : m_name(name), has_been_set(false) {}
+public:
+	void set(const T value[count], video::IMaterialRendererServices *services)
+	{
+		if (std::equal(m_sent, m_sent + count, value) && has_been_set)
```

Comment:
While it is much cheaper, it's true almost 100% of the time. std::equal is false far more often. There is another (good) reason to flip the order however.